src/services/download-service.ts

```diff
@@ -12,123 +13,153 @@ class DownloadService {
     return !(this.player.isLive() || this.player.getVideoElement().mediaKeys);
   }
 
-  private isContentTypeSupported(response: Response) {
-    if (response.ok) {
-      const contentType = response.headers.get('content-type');
-      return !!(contentType && contentType.toLowerCase().includes('video'));
-    }
-    return false;
-  }
-
-  private getFilename(response: Response) {
-    if (this.player.isImage()) {
-      return this.player.sources?.metadata?.name || 'image';
-    }
-    const responseUrlSplit = response.url.split('/');
+  private getFilename(metadata: DownloadMetadata) {
+    if (this.player.sources.metadata?.name) return this.player.sources.metadata?.name;
+    if (this.player.isImage()) return 'image';
+    const responseUrlSplit = metadata!.flavors[0].downloadUrl.split('/');
     return responseUrlSplit[responseUrlSplit.indexOf('fileName') + 1];
   }
 
-  private getDownloadUrl(config: DownloadConfig) {
-    const {flavorId, flavorParamId} = config;
-    const {provider} = this.player;
-
-    const playerConfig = this.player.config;
-    const {session, sources} = playerConfig;
-
-    if (!(session && session.ks && sources && sources.id && provider.partnerId && provider.env.cdnUrl)) {
-      return '';
+  async getDownloadMetadata(): Promise<DownloadMetadata> {
+    if (!(this.isPlatformSupported() && this.isEntrySupported())) {
+      return null;
     }
 
-    const {ks} = playerConfig.session;
-    const {id} = playerConfig.sources;
-    const {partnerId} = provider;
-    const {cdnUrl} = provider.env;
+    const metadata: DownloadMetadata = {
+      fileName: '',
+      captions: [],
+      flavors: [],
+      attachments: [],
+      imageDownloadUrl: ''
+    };
+
+    const assets = await this.getKalturaAssets();
+    Object.assign(metadata, assets);
+    metadata!.imageDownloadUrl = await this.handleImageDownload();
+    const downloadUrls: Map<string, string> = await this.getDownloadUrls(metadata);
+
+    // assign the download urls to the assets by id
+    metadata?.captions.forEach(captionAsset => (captionAsset.downloadUrl = downloadUrls.get(captionAsset.id)!));
+    metadata?.flavors.forEach(flavorAsset => (flavorAsset.downloadUrl = downloadUrls.get(flavorAsset.id)!));
+    metadata?.attachments.forEach(attachmentAsset => (attachmentAsset.downloadUrl = downloadUrls.get(attachmentAsset.id)!));
```

Comment:
you can also use .map instead of .foreach here i think just to make it more readable. unless you think it makes it *less* readable